cava/samples/cudart/cudart.cpp

```diff
@@ -435,21 +437,51 @@ cudaFreeHost(void *ptr)
 }
 ava_end_replacement;
 
+
+/// Migration: @mem_extract_tag
+ava_utility void* object_extract(void* obj, size_t* length) {
+    //called from host
+    void* buffer;
+    printf("object_replay: object=%lx\n", (uintptr_t)obj);
+    cudaDeviceSynchronize();
+
+    *length = ava_metadata(obj)->buffer_size;
+    buffer = malloc(*length);
+    cudaMemcpy(buffer, obj, *length, cudaMemcpyDeviceToHost);
+    return buffer;
+}
+
+ava_utility void object_replace(void* obj, void* data, size_t length) {
+    printf("object_replace: object=%lx, len=%lu\n", (uintptr_t)obj, length);
+    assert(length != 0);
+    cudaMemcpy(obj, data, length, cudaMemcpyHostToDevice);
+}
+
 __host__ __cudart_builtin__ cudaError_t CUDARTAPI
 cudaMalloc(void **devPtr, size_t size)
 {
     ava_argument(devPtr) {
+
         ava_out; ava_buffer(1);
-        ava_element ava_opaque;
+        ava_element{
+            ava_allocates;
+            ava_handle;
+            ava_object_explicit_state_functions(object_extract, object_replace);
+            ava_object_record;
+        }
     }
+
+    ava_execute();
+    ava_metadata(*devPtr)->buffer_size = size;
 }
 
 __host__ cudaError_t CUDARTAPI
 cudaMemcpy(void *dst, const void *src, size_t count, enum cudaMemcpyKind kind)
 {
     ava_argument(dst) {
         if (kind == cudaMemcpyHostToDevice) {
-            ava_opaque;
+            ava_handle;
```

Comment:
As this change requires to annotate devptr as handles which doesn't support the "offsetting" (+/-) operation I think we should merge this after we redesign the handle.

Reply:
Do you know of any test applications that this breaks?

Reply:
There is another issue. Devptrs should be handles everywhere or nowhere. I don't think this spec makes the devptrs in arguments to kernels handles.

Reply:
> Do you know of any test applications that this breaks?

Not for Rodinia CUDA benchmarks. But for the supported AI frameworks (like TF), copying memory at offsets is a normal behavior.

Reply:
> 
> 
> There is another issue. Devptrs should be handles everywhere or nowhere. I don't think this spec makes the devptrs in arguments to kernels handles.

@arthurp The spec for cuLaunchKernel already checked if the argument was a handle and applied the annotation accordingly. Is this not doing the right thing?

```
CUresult CUDAAPI
cuLaunchKernel(CUfunction f,
               unsigned int gridDimX,
               unsigned int gridDimY,
               unsigned int gridDimZ,
               unsigned int blockDimX,
               unsigned int blockDimY,
               unsigned int blockDimZ,
               unsigned int sharedMemBytes,
               CUstream hStream,
               void **kernelParams,
               void **extra)
{
    ava_argument(hStream) ava_handle;

    ava_argument(kernelParams) {
        ava_in; ava_buffer(ava_metadata(f)->func->argc);
        ava_element {
            // FIXME: use the generated index name in the spec to
            // reference the outer loop's loop index at this moment.
            if (ava_metadata(f)->func->args[__kernelParams_index_0].is_handle) {
                ava_type_cast(void *);
                ava_buffer(ava_metadata(f)->func->args[__kernelParams_index_0].size);
                ava_element ava_handle;
            }
            else {
                ava_type_cast(void *);
                ava_buffer(ava_metadata(f)->func->args[__kernelParams_index_0].size);
            }
        }
    }

    ava_argument(extra) {
        ava_in; ava_buffer(__helper_launch_extra_size(extra));
#warning The buffer size below states that every kernelParams[i] is 1 byte long.
        ava_element ava_buffer(1);
    }
}
```

Reply:
A mergeSort cuda sample is failing with cudaErrorMisalignedAddress, which seems related to handle->devPtr translation. Are there other parts of the spec that need the handle annotation?

Reply:
It depends on how the application uses those handles (to them they are devPtr). If they add offsets to those CUDA "addresses", it's incorrect to annotate them as handles.
I'm not sure but `cudaErrorMisalignedAddress` is very likely to be caused by this issue.

Reply:
Seeing the same `CUDA Runtime API error 716: misaligned address.` on other benchmarks. These benchmarks do not do pointer arithmetic on the host side.

Reply:
How does the ava_argument(args) code from the cuda spec change the generated guestlib/worker code?

It appears that the worker calls out to `__helper_launch_kernel` in `cudart_10.1_utilities.cpp`. That function passes `args` unmodified to the native `cudaLaunchKernel` call, so the worker should be translating any handles in `args` to device pointers. I don't see that happening.

There is some argument processing in the generated `cudart_nw_worker.cpp`  but I don't think it translates handles to device pointers.

```
/* Input: void ** args */
        void **args; {
            args =
                (void **)(((__call->args) != (NULL)) ? ((void **)command_channel_get_buffer(__chan, __cmd,
                        __call->args)) : ((void **)__call->args));
            if ((__call->args) != (NULL)) {
                void **__src_args_0;
                __src_args_0 = (void **)(args);
                volatile size_t __buffer_size = 0;
                __buffer_size = ((size_t) (ava_metadata(func)->func->argc));
                args = (void **)((void **)command_channel_get_buffer(__chan, __cmd, __call->args));
                if ((__call->args) != (NULL)) { {
                        const size_t __size = ((size_t) (ava_metadata(func)->func->argc));
                        args = (void **)(malloc(__size * sizeof(void *)));
                        g_ptr_array_add(__ava_alloc_list_cudaLaunchKernel, ava_buffer_with_deallocator_new(free, args));
                }}

                const size_t __args_size_0 = __buffer_size;
                for (size_t __args_index_0 = 0; __args_index_0 < __args_size_0; __args_index_0++) {
                    const size_t ava_index = __args_index_0;

                    void **__args_a_0;
                    __args_a_0 = (void **)(args) + __args_index_0;

                    void **__args_b_0;
                    __args_b_0 = (void **)(__src_args_0) + __args_index_0;

                    *__args_a_0 =
                        (void *)(((*__args_b_0) != (NULL)) ? ((void *)command_channel_get_buffer(__chan, __cmd,
                                *__args_b_0)) : ((void *)*__args_b_0)); {
                        *__args_a_0 =
                            (void *)(((*__args_b_0) != (NULL)) ? ((void *)command_channel_get_buffer(__chan, __cmd,
                                    *__args_b_0)) : ((void *)*__args_b_0));
                        if ((*__args_b_0) != (NULL)) {
                            void *__src_args_1;
                            __src_args_1 = (void *)(*__args_a_0);
                            volatile size_t __buffer_size = 0;
                            __buffer_size = ((size_t) (ava_metadata(func)->func->args[__args_index_0].size));
                            *__args_a_0 = (void *)((void *)command_channel_get_buffer(__chan, __cmd, *__args_b_0));

                            if ((*__args_a_0) != (__src_args_1)) {
                                memcpy(*__args_a_0, __src_args_1, __buffer_size * 1);
                            }
                        } else {
                            *__args_a_0 =
                                (void *)(((*__args_b_0) != (NULL)) ? ((void *)command_channel_get_buffer(__chan, __cmd,
                                        *__args_b_0)) : ((void *)*__args_b_0));
                    }}
            }} else {
                if ((__call->args) != (NULL)) { {
                        const size_t __size = ((size_t) (ava_metadata(func)->func->argc));
                        args = (void **)(malloc(__size * sizeof(void *)));
                        g_ptr_array_add(__ava_alloc_list_cudaLaunchKernel, ava_buffer_with_deallocator_new(free, args));
                }}
        }}
```

Reply:
If you haven't touched that piece of spec, then it treats devPtr kernel args as non-handle.

Reply:
What should it look like? It seems like it checks if each argument is a handle and if it is adds the ava_handle annotation. Is this doing something different?

cudaLaunchKernel spec:
```
 ava_argument(args) {
    ava_in;
    ava_buffer(ava_metadata(func)->func->argc);
    ava_element {
      // FIXME: use the generated index name in the spec to
      // reference the outer loop's loop index at this moment.
      if (ava_metadata(func)->func->args[__args_index_0].is_handle) {
        ava_type_cast(void *);
        ava_buffer(ava_metadata(func)->func->args[__args_index_0].size);
        // ava_element ava_handle;
      } else {
        ava_type_cast(void *);
        ava_buffer(ava_metadata(func)->func->args[__args_index_0].size);
      }
    }
    ```

Reply:
oh, it's commented out now. Should it be uncommented?

Reply:
Uncommented `ava_element ava_handle` and the behavior is unchanged.

Reply:
> 
> 
> If you haven't touched that piece of spec, then it treats devPtr kernel args as non-handle.

@yuhc can you elaborate on this?